Documentation/design-docs/raw-eventlistener.md

```diff
@@ -0,0 +1,81 @@
+# Raw EventListener API #
+
+The goal of this design document is to describe what a Raw EventListener API will look like and how it will work.
+
+A raw API is one that dispatches events, but does not decode the data payload.  It simply provides a raw blob and a metadata blob that can be used to interpret the data when desired.
+
+## API Surface ##
+
+The existing EventListener API usage pattern is to:
+
+1. Create a new class that derives from EventListener.
+2. Implement the OnEventSourceCreated method to get callbacks when new EventSource objects are created (this allows EventListeners to subscribe to events).
+3. Implement the OnEventWritten method to get callbacks whenever a subscribed event is dispatched.
+
+The proposed extension of this pattern is the following:
+
+1. Add a new EventListener constructor that takes an EventListenerSettings enum parameter:
+
+```
+[Flags]
+public enum EventListenerSettings
+{
+   None,
+   RawEventDispatch
+}
+``` 
+
+This parameter is used to specify the desired dispatch behavior (in this case, do not deserialize event payloads).
+
+2. Depending on the configuration, EventListener will call the traditional (deserialized payload) API or the raw API.
+
+The new raw dispatch API will be:
+
+```
+public void OnEventWrittenRaw(RawEventWrittenEventArgs args);
+ 
+public sealed class RawEventWrittenEventArgs
+{
+
+    // Event metadata copied from EventWrittenEventArgs (for consistency).
+    public string EventName { get; }
+    public int EventId { get; }
+    public Guid ActivityId { get; }
+    public Guid RelatedActivityId { get; }
+    public EventSource EventSource { get; }
+    public EventKeywords Keywords { get; }
+    public EventOpcode Opcode { get; }
+    public EventTask Task { get; }
+    public EventTags Tags { get; }
+    public string Message { get; }
+    public byte Version { get; }
+    public EventLevel Level { get; }
+    public long OSThreadId { get; }
+    public DateTime TimeStamp { get; }
+    
+    // Replacement properties for Payload and PayloadNames.
+    public ReadOnlySpan<byte> Metadata { get; }     
+    public ReadOnlySpan<byte> Payload { get; }
```

Comment:
Could the runtime provide "metadata version" integer in this event that would change any time the Metadata or Payload formats change? We can tag a version number on the payload before sending it out of process but I'd prefer if that version were controlled by the runtime since it's the one generating the blobs.

A sample int (starting at `1` here) would be sufficient, we'd consider any change in version "breaking" and update consumption tools to handle it and we wouldn't expect any back-compat or for the runtime to support older versions. As long as it bumps whenever a change is made that couldn't be decoded by a parser built for the current format.

Reply:
A second unrelated question: Do these blobs include all the above information (`Keywords`, `Opcodes`, etc.) as well? Which properties would a "shipper" need to "ship" to be able to recreate the event?

Reply:
I don't think we have the ability to produce a metadata version without manual intervention from the user.  The reason is that the metadata is produced by serializing data such as event ID, payload argument information, etc. and this data can only change by the code being modified.  As such, when the process boots it doesn't know that there was a previous version of the metadata.  What you can be guaranteed is that for a given process, the metadata for an event won't change.

Reply:
The metadata blob contains:
 - Provider Name
 - Event ID
 - Event Name
 - Keywords
 - Version
 - Level

Opcode is inferred by the eventName (we basically support Start, Stop and Info).

Reply:
I think you misunderstand what I mean by "metadata version" :). I mean a version number defining **how** the metadata is encoded. So I guess a "serialization format version" is more accurate. I realize there's no way for the runtime to know if the event definition has changed. I'm looking for a version number, stored next to the serialization logic, that can signal when the *serialization logic* has changed (if, say, we suddenly decide ProtoBuf or MessagePack are better formats, or decide to change a field from an int32 to an int64, etc.).

Reply:
Ah.  Gotcha.  This makes sense.  Let me think about this and I'll update the doc.